### PR TITLE
Safer Cluster module

### PR DESCRIPTION
**Pull request by bharathkkb**
- Added commits from #272 and docs #281 
- Resolved comments from #272 
- Added new build step, shielded vms

### CODE REVIEW

File: modules/safer-cluster/main.tf [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  initial_node_count = var.initial_node_count
+
+  // We suggest removing the default node pull, as it cannot be modified without
+  // destroying the cluster.
+  remove_default_node_pool = true
+
+  node_pools        = var.node_pools
+  node_pools_labels = var.node_pools_labels
+
+  // TODO(mmontan): check whether we need to restrict these
```
Comment (aaron-lane) [inline modules/safer-cluster/main.tf]: Let's replace this comment with an issue.

File: modules/safer-cluster/main.tf [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  // settings.
+  node_pools_metadata = var.node_pools_metadata
+  node_pools_taints   = var.node_pools_taints
+  node_pools_tags     = var.node_pools_tags
+
+  // TODO(mmontan): we generally considered applying
```
Comment (aaron-lane) [inline modules/safer-cluster/main.tf]: Let's replace this comment with an issue.

File: modules/safer-cluster/main.tf [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  // We never use the default service account for the cluster. The default
+  // project/editor permissions can create problems if nodes were to be ever
+  // compromised.
+
+  // We either:
+  // - Create a dedicated service account with minimal permissions to run nodes.
+  //   All applications shuold run with an identity defined via Workload Identity anyway.
+  // - Use a service account passed as a parameter to the module, in case the user
+  //   wants to maintain control of their service accounts.
+  create_service_account = var.compute_engine_service_account == "" ? false : true
```
Comment (aaron-lane) [inline modules/safer-cluster/main.tf]: This logic appears to be inverted; `create_service_account` should be `true` if `compute_engine_service_account` is empty.

File: modules/safer-cluster/main.tf [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+  service_account        = var.compute_engine_service_account
+
+  // TODO(mmontan): define a registry_project parameter in the private_beta_cluster,
```
Comment (aaron-lane) [inline modules/safer-cluster/main.tf]: Let's replace this comment with an issue.

File: modules/safer-cluster/main.tf [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+  default_max_pods_per_node = var.default_max_pods_per_node
+
+  database_encryption = var.database_encryption
+
+  // We suggest to define policies about  which images can run on a cluster.
+  enable_binary_authorization = true
+
+  // Define PodSecurityPolicies for differnet applications.
+  // TODO(mmontan): link to a couple of policies.
```
Comment (aaron-lane) [inline modules/safer-cluster/main.tf]: Let's replace this comment with an issue.

File: modules/safer-cluster/main.tf [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    "enabled" = true
+  }]
+
+  resource_usage_export_dataset_id = var.resource_usage_export_dataset_id
+
+  // Sandbox is needed if the cluster is going to run any untrusted workload (e.g., user submitted code).
+  // Sandbox can also provide increased protection in other cases, at some performance cost.
+  sandbox_enabled = var.sandbox_enabled
+
+  // TODO(mmontan): investigate whether this should be a recommended setting
```
Comment (aaron-lane) [inline modules/safer-cluster/main.tf]: Let's replace this comment with an issue.

Comment (bharathkkb) [reply modules/safer-cluster/main.tf]: This is a recommended best practice, so removing the todo

File: modules/safer-cluster/main.tf (hunk 5 above)
Comment (bharathkkb) [reply modules/safer-cluster/main.tf]: Linked an example policy from kubernetes.io.

File: modules/safer-cluster/main.tf (hunk 4 above)
Comment (bharathkkb) [reply modules/safer-cluster/main.tf]: Added registry_project_id

File: modules/safer-cluster/main.tf (hunk 3 above)
Comment (bharathkkb) [reply modules/safer-cluster/main.tf]: Fixed

File: modules/safer-cluster/main.tf (hunk 2 above)
Comment (bharathkkb) [reply modules/safer-cluster/main.tf]: Tracked at #319

File: modules/safer-cluster/main.tf (hunk 1 above)
Comment (bharathkkb) [reply modules/safer-cluster/main.tf]: tracked at #320

File: modules/safer-cluster/README.md [hunk 7]
```diff
@@ -0,0 +1,280 @@
+# Safer Cluster: How to setup a GKE Kubernetes cluster with reduced exposure
+
+[TOC]
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply modules/safer-cluster/README.md]: We dont seem to have a TOC for other modules; so removing this.

File: test/fixtures/safer_cluster/network.tf [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  length  = 4
+  special = false
+  upper   = false
+}
+
+provider "google" {
+  project = var.project_id
+}
+
+resource "google_compute_network" "main" {
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply test/fixtures/safer_cluster/network.tf]: Done

File: test/ci/safer-cluster.yml [hunk 9]
```diff
@@ -0,0 +1,18 @@
+---
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply test/ci/safer-cluster.yml]: Done

File: examples/safer_cluster/main.tf [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+ * limitations under the License.
+ */
+
+locals {
+  cluster_type = "safer-cluster"
+}
+
+provider "google-beta" {
+  version = "~> 2.18.0"
+  region  = var.region
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply examples/safer_cluster/main.tf]: Done

File: examples/safer_cluster/variables.tf [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+variable "project_id" {
+  description = "The project ID to host the cluster in"
```
Comment (aaron-lane) [inline examples/safer_cluster/variables.tf]: Variables should be typed.

File: modules/safer-cluster/main.tf (hunk 1 above)
Comment (aaron-lane) [reply modules/safer-cluster/main.tf]: The comment can be removed.

File: modules/safer-cluster/variables.tf [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// This file was automatically generated from a template in ./autogen
+
+variable "project_id" {
+  type        = string
+  description = "The project ID to host the cluster in (required)"
```
Comment (aaron-lane) [inline modules/safer-cluster/variables.tf]: Variable descriptions do not need to indicate the optional or required status unless the usage is complex.

File: .kitchen.yml [hunk 13]
```diff
@@ -45,6 +45,13 @@ suites:
       systems:
         - name: shared_vpc
           backend: local
+  - name: "safer_cluster"
```
[the remark this answers is not in the document]
Comment (morgante) [reply .kitchen.yml]: I'm fine with making this a separate suite, as it's distinct enough.. I'd rather focus our consolidation on existing ones.

File: examples/safer_cluster/main.tf [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+locals {
+  cluster_type = "safer-cluster"
+}
+
+provider "google-beta" {
+  version = "~> 2.18.0"
+}
+
+data "google_compute_subnetwork" "subnetwork" {
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply examples/safer_cluster/main.tf]: Done

File: examples/safer_cluster/variables.tf (hunk 11 above)
Comment (bharathkkb) [reply examples/safer_cluster/variables.tf]: Added types

File: modules/safer-cluster/variables.tf (hunk 12 above)
Comment (bharathkkb) [reply modules/safer-cluster/variables.tf]: fixed

**Comment by bharathkkb** [comment]
@aaron-lane seems like it timed out at destroy sandbox-enabled-local

File: examples/safer_cluster/variables.tf [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+  default     = ""
+}
+
+variable "region" {
+  type        = string
+  description = "The region to host the cluster in"
+  default     = "us-central1"
+}
+
+variable "network" {
```
Comment (aaron-lane) [inline examples/safer_cluster/variables.tf]: I recommend removing most of these variables and replacing them with hard-coded values in the example. Additionally, any names should include a random component from a resource like `random_pet_name`.

Comment (bharathkkb) [reply examples/safer_cluster/variables.tf]: Done

**Pull request merged**